src/rt/minfo.d

```diff
@@ -60,6 +60,339 @@ struct ModuleGroup
+                reachable[] = 0;
+                foreach(e; m.importedModules)
+                {
+                    auto impidx = findModule(e);
+                    if(impidx != -1 && impidx != i)
+                        reachable[impidx] = 1;
+                }
+
+                foreach(int j, r; reachable)
+                {
+                    if(r)
+                        edges[i] ~= j;
+                }
+            }
+
+            // a cycle starts with the source.
+            cyclePath ~= srcidx;
+            // first get to the target
+            shortest(srcidx, targetidx);
+            // now get back.
+            shortest(targetidx, srcidx);
+        }
+
+        // find all the non-trivial dependencies (that is, dependencies that have a
+        // ctor or dtor) of a given module.  Doing this, we can 'skip over' the
+        // trivial modules to get at the non-trivial ones.
+        size_t _findDependencies(int idx, bool orig = true)
+        {
+            if(reachable[idx])
+                return 0;
+            auto current = _modules[idx];
+            size_t result = 0;
+            reachable[idx] = 1;
+            if(!orig && (flags[idx] & (MIctor | MIdtor)) && !(flags[idx] & MIstandalone))
+                // non-trivial, stop here
+                return result + 1;
+            foreach (m; current.importedModules)
+            {
+                auto midx = findModule(m);
+                if(midx != -1)
+                    // not part of this DSO, don't consider it.
+                    result += _findDependencies(midx, false);
+            }
+            return result;
+        }
+
+        // This function will determine the order of construction/destruction and
+        // check for cycles. If a cycle is found, the cycle path is transformed
+        // into a string and thrown as an error.
+        //
+        // Each call into this function is given a module that has static
+        // ctor/dtors that must be dealt with. It recurses only when it finds
+        // dependencies that also have static ctor/dtors.
+        void _checkModCtors2(int curidx)
+        {
+            assert(curidx != -1);
+            immutable ModuleInfo* current = _modules[curidx];
+
+            // we only get called if current has a dtor or a ctor, so no need to
+            // check that.  First, determine what non-trivial elements are
+            // reachable.
+            reachable[] = 0;
+            auto nmodules = _findDependencies(curidx);
+
+            // allocate the dependencies on the stack
+            auto p = cast(int *)alloca(nmodules * int.sizeof);
+            auto dependencies = p[0 .. nmodules];
+            uint depidx = 0;
+            // fill in the dependencies
+            foreach (int i, r; reachable)
+            {
+                if(r && i != curidx)
+                {
+                    if((flags[i] & (MIctor | MIdtor)) && !(flags[i] & MIstandalone))
```

Comment:
This could already be filtered out during reachability computation, no need for a second array.
We could just use `auto reachable = findDependencies(current)` and let it return a bit vector (Array!ubyte or so).